addons/binding/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/handler/TeslaHandler.java

```diff
@@ -0,0 +1,948 @@
+				// object for some specific non-JSON compatible requests
+				switch (request) {
+				case TESLA_MOBILE_ENABLED_STATE: {
+					jsonObject = new JsonObject();
+					jsonObject.addProperty(TESLA_MOBILE_ENABLED_STATE, result);
+					break;
+				}
+				default: {
+					jsonObject = parser.parse(result).getAsJsonObject();
+					break;
+				}
+				}
+			}
+
+			// process the result
+			if (jsonObject != null && result != null && !result.equals("null")) {
+				// deal with responses for "set" commands, which get confirmed
+				// positively, or negatively, in which case a reason for failure
+				// is provided
+				if (jsonObject.get("reason") != null
+						&& jsonObject.get("reason").getAsString() != null) {
+					boolean requestResult = jsonObject.get("result")
+							.getAsBoolean();
+					logger.debug(
+							"The request ({}) execution was {}, and reported '{}'",
+							new Object[] {
+									request,
+									requestResult ? "successful"
+											: "not successful",
+									jsonObject.get("reason").getAsString() });
+				} else {
+					Set<Map.Entry<String, JsonElement>> entrySet = jsonObject
+							.entrySet();
+					for (Map.Entry<String, JsonElement> entry : entrySet) {
+						try {
+							TeslaChannelSelector selector = TeslaChannelSelector
+									.getValueSelectorFromRESTID(entry.getKey());
+							if (!selector.isProperty()) {
+								ChannelUID theChannelUID = new ChannelUID(
+										getThing().getUID(),
+										selector.getChannelID());
+								if (!entry.getValue().isJsonNull()) {
+									updateState(theChannelUID,
+											selector.getState(entry.getValue()
+													.getAsString()));
+								} else {
+									updateState(theChannelUID, UnDefType.UNDEF);
+								}
+							} else {
+								if (!entry.getValue().isJsonNull()) {
+									Map<String, String> properties = editProperties();
+									properties.put(selector.getChannelID(),
+											entry.getValue().getAsString());
+									updateProperties(properties);
+								}
+							}
+						} catch (Exception e) {
+							logger.debug(
+									"Unable to handle the variable/value pair '{}':'{}'",
+									entry.getKey(), entry.getValue());
+						}
+					}
+				}
+			}
+		} catch (Exception p) {
+			logger.error(
+					"An exception occurred while parsing data received from the vehicle: '{}'",
+					p.getMessage());
+			p.printStackTrace();
+		}
+
+	}
+
+	protected boolean isAwake() {
+		return (vehicle != null) ? (vehicle.state != "asleep" && vehicle.vehicle_id != null)
+				: false;
+	}
+
+	protected boolean isInMotion() {
+		if (driveState != null) {
+			if (driveState.speed != null && driveState.shift_state != null) {
+				return !driveState.speed.equals("Undefined")
+						&& (!driveState.shift_state.equals("P") || !driveState.shift_state
+								.equals("Undefined"));
+			}
+		}
+		return false;
+	}
+
+	public void setChargeLimit(int percent) {
+		Gson gson = new Gson();
```

Comment:
this is a potentially expensive operation. Better only instantiate this once as a field and use it from the different places within this class where you need it.